apps/webapp/src/state/swap.ts

```diff
@@ -1,103 +1,105 @@
-    setAsset: type => asset => {
-      let payAsset = get().swap.pay.asset;
-      let receiveAsset = get().swap.receive.asset;
-
-      const selectedAsset = assets.find(i => i.penumbraAssetId.inner === asset.inner)!;
-
-      // checking if we set the same asset in the pay (receive) field as
-      // in the receive (pay) field, then the value of the receive (pay)
-      // field is undefined
-      if (type === SwapInputs.PAY) {
-        if (selectedAsset.display === receiveAsset.display) {
-          receiveAsset = assets.find(i => i.display !== selectedAsset.display)!;
-        }
-        payAsset = selectedAsset;
-      } else {
-        if (selectedAsset.display === payAsset.display) {
-          payAsset = assets.find(i => i.display !== selectedAsset.display)!;
-        }
-        receiveAsset = selectedAsset;
-      }
-
+    assetOut: undefined,
+    setAssetOut: denom => {
       set(state => {
-        state.swap.pay.asset = payAsset;
-        state.swap.receive.asset = receiveAsset;
+        state.swap.assetOut = denom;
       });
     },
-    setAssetBalance: () => {
-      // TODO: Finish later
-      // const { amount } = get().swap.pay;
-      // set(state => {
-      // state.swap.pay.balance = balance;
-      // state.swap.validationErrors.pay = validateAmount(amount, balance);
-      // });
+    amount: '',
+    setAmount: amount => {
+      set(state => {
+        state.swap.amount = amount;
+      });
     },
-    replaceAsset: () => {
-      const pay = get().swap.pay;
-      const receive = get().swap.receive;
-
+    txInProgress: false,
+    initiateSwapTx: async toastFn => {
       set(state => {
-        state.swap.pay = { ...receive, balance: 0 };
-        state.swap.receive = { amount: pay.amount, asset: pay.asset };
+        state.swap.txInProgress = true;
       });
+
+      const { dismiss } = toastFn(loadingTxToast);
+
+      try {
+        const req = await assembleRequest(get().swap);
+        const txHash = await planWitnessBuildBroadcast(req);
+        dismiss();
+        toastFn(successTxToast(txHash));
+
+        // Reset form
+        set(state => {
+          state.send.amount = '';
+          state.send.txInProgress = false;
+        });
+      } catch (e) {
+        set(state => {
+          state.send.txInProgress = false;
+        });
+        dismiss();
+        toastFn(errorTxToast(e));
+        throw e;
+      }
     },
   };
 };
 
+const assembleRequest = async ({ assetIn, amount, assetOut }: SwapSlice) => {
+  if (assetIn?.accountIndex === undefined || !assetIn.asset || !assetOut?.metadata.penumbraAssetId)
+    throw new Error('missing selected tokens');
+
+  return new TransactionPlannerRequest({
+    swaps: [
+      {
+        targetAsset: assetOut.metadata.penumbraAssetId,
+        value: {
+          amount: toBaseUnit(BigNumber(amount), assetIn.asset.denom.exponent),
+          assetId: assetIn.asset.assetId,
+        },
+        claimAddress: await getEphemeralAddress(assetIn.accountIndex),
```

Comment:
Why an ephemeral address? Is that just a default we want to use? (Seems unnecessary in a swap though, since no one else can read anything in the transaction, right?)

Reply:
we want to avoid using one-time addresses anywhere except ibc deposits, where they're absolutely necessary, this should indeed be using the default address for the account